vignettes/corrected-se.Rmd

```diff
@@ -0,0 +1,322 @@
+vldev2 <- attr(fs2, which = "vfsLT")
+fs_dat <- data.frame(
+  fs_ind60 = fs1$fs_ind60,
+  fs_dem60 = fs2$fs_dem60
+)
+# Combine sampling variance of loading and error variance
+# Note: loadings first, then error variance
+vldev <- block_diag(vldev1, vldev2)[c(1, 3, 2, 4), c(1, 3, 2, 4)]
+```
+
+```{r}
+# 2S-PA
+# Assemble loadings
+ld <- block_diag(attr(fs1, which = "fsL"),
+                 attr(fs2, which = "fsL"))
+ev <- block_diag(attr(fs1, which = "fsT"),
+                 attr(fs2, which = "fsT"))
+tspa_fit <- tspa(model = "dem60 ~ ind60",
+                 data = fs_dat,
+                 cross_loadings = ld,
+                 vc = ev)
+# Unadjusted covariance
+vcov(tspa_fit)
+# Adjusted covariance matrix
+(vc_cor <- vcov_corrected(tspa_fit, vfsLT = vldev, which_free = c(1, 4, 5, 7)))
+# Corrected standard errors
+sqrt(diag(vc_cor))
+```
+
+Standardized solution
+
+```{r}
+tspa_est_std <- function(ld_ev) {
+  ld1 <- ld
+  ev1 <- ev
+  ld1[c(1, 4)] <- ld_ev[1:2]
+  ev1[c(1, 4)] <- ld_ev[3:4]
+  tfit <- tspa(model = "dem60 ~ ind60",
+               data = fs_dat,
+               cross_loadings = ld1,
+               vc = ev1)
+  standardizedSolution(tfit)$est.std[8]
+}
+tspa_est_std(c(ld[c(1, 4)], ev[c(1, 4)]))
+# Jacobian
+jac_std <- numDeriv::jacobian(tspa_est_std,
+                              x = c(ld[c(1, 4)], ev[c(1, 4)]))
+# Corrected standard error
+sqrt(
+  standardizedSolution(tspa_fit)[8, "se"]^2 +
+    jac_std %*% vldev %*% t(jac_std)
+)
+```
+
+Compare to joint model
+
+```{r, eval = FALSE}
+sem_fit <- sem(model, data = PoliticalDemocracy)
+# Larger standard error
+vcov(sem_fit)[c("dem60~ind60", "ind60~~ind60", "dem60~~dem60"),
+              c("dem60~ind60", "ind60~~ind60", "dem60~~dem60")]
+```
+
+Bootstrap Standard Errors
+
+```{r, eval = FALSE, echo = TRUE}
+run_tspa <- function(df, inds) {
+  fs_ind60 <- get_fs(data = df[inds, ], model = "ind60 =~ x1 + x2 + x3",
+                     se = "none", test = "none")
+  fs_dem60 <- get_fs(data = df[inds, ], model = "dem60 =~ y1 + y2 + y3 + y4",
+                     se = "none", test = "none")
+  fs_dat <- cbind(fs_ind60, fs_dem60)
+  # Assemble loadings
+  ld <- block_diag(attr(fs_ind60, which = "fsL"),
+                   attr(fs_dem60, which = "fsL"))
+  ev <- block_diag(attr(fs_ind60, which = "fsT"),
+                   attr(fs_dem60, which = "fsT"))
+  tspa_fit <- tspa(model = "dem60 ~ ind60",
+                   data = fs_dat,
+                   cross_loadings = ld,
+                   vc = ev,
+                   test = "none")
+  coef(tspa_fit)
+}
+boo <- boot(PoliticalDemocracy, statistic = run_tspa, R = 1999)
+```
+
+```{r, eval = FALSE, include = FALSE}
+saveRDS(boo, file = "boo_separate.RDS")
+```
+
+```{r, include = FALSE}
+boo <- readRDS("boo_separate.RDS")
+```
+
+```{r}
+# Use MAD to downweigh outlying replications
+boo$t |>
+    apply(MARGIN = 2, FUN = mad) |>
+    setNames(c("dem60~ind60", "ind60~~ind60", "dem60~~dem60"))
```

Comment:
Bootstrap
```
> boo$t |>
+     apply(MARGIN = 2, FUN = mad) |>
+     setNames(c("dem60~ind60", "ind60~~ind60", "dem60~~dem60"))
 dem60~ind60 ind60~~ind60 dem60~~dem60 
  0.32521386   0.07271009   0.87676911 
```
Joint model
```
> sqrt(diag(vcov(sem_fit)[c("dem60~ind60", "ind60~~ind60", "dem60~~dem60"),
+                    c("dem60~ind60", "ind60~~ind60", "dem60~~dem60")]))
 dem60~ind60 ind60~~ind60 dem60~~dem60 
  0.37862348   0.08675722   0.87496855 
```
Corrected 2S-PA
```
> sqrt(diag(vc_cor))
 dem60~ind60 ind60~~ind60 dem60~~dem60 
  0.36415174   0.07585532   0.71996886 
  ```
The standard errors seem to diverge slightly among methods. SE(`dem60~ind60`) with bootstrap is the lowest. SE(`ind60~~ind60`) with the joint model is particularly higher than the other two. SE(`dem60~~dem60`) is the lowest with the corrected 2S-PA.

Reply:
Will add this to the vignette